File: backend/app/schemas/validation.py
"""Pydantic schemas for Vision Performance Validation module."""
from __future__ import annotations
from typing import Optional
from pydantic import BaseModel


class GRRInput(BaseModel):
    values: list[list[list[float]]]  # [operator][part][replicate]
    n_operators: int = 2
    n_parts: int = 10
    n_replicates: int = 2
    tolerance: Optional[float] = None


class ValidationRequest(BaseModel):
    system_name: str = "Vision System"
    # Repeatability / Measurement
    measurements: Optional[list[float]] = None
    reference_value: Optional[float] = None
    usl: Optional[float] = None
    lsl: Optional[float] = None
    reference_values: Optional[list[float]] = None   # for accuracy (paired)
    # AI Performance
    predictions: Optional[list[int]] = None
    ground_truth: Optional[list[int]] = None
    scores: Optional[list[float]] = None             # probability scores for ROC
    class_names: Optional[list[str]] = None
    # OCR
    ocr_predicted: Optional[list[str]] = None
    ocr_ground_truth: Optional[list[str]] = None
    ocr_confidences: Optional[list[float]] = None
    # Runtime
    inference_times_ms: Optional[list[float]] = None
    target_fps: Optional[float] = None
    # Stability
    stability_results: Optional[list[bool]] = None
    # GR&R
    grr: Optional[GRRInput] = None


# ── Result schemas ─────────────────────────────────────────────────────────────

class RepeatabilityResult(BaseModel):
    n: int
    mean: float
    std: float
    min_val: float
    max_val: float
    range_val: float
    cp: Optional[float]
    cpk: Optional[float]
    cv_pct: float
    usl: Optional[float]
    lsl: Optional[float]
    histogram: list[float]
    histogram_edges: list[float]
    status: str   # capable / marginal / incapable


class AccuracyResult(BaseModel):
    n: int
    mae: float
    mape_pct: float
    bias: float
    rmse: float
    max_error: float
    errors: list[float]
    status: str


class GRRResult(BaseModel):
    n_operators: int
    n_parts: int
    n_replicates: int
    ev_pct: float
    av_pct: float
    grr_pct: float
    pv_pct: float
    ndc: int
    var_repeatability: float
    var_reproducibility: float
    var_parts: float
    status: str   # acceptable / marginal / unacceptable


class AIResult(BaseModel):
    n_samples: int
    accuracy: float
    precision: float
    recall: float
    f1_score: float
    confusion_matrix: list[list[int]]
    class_names: list[str]
    per_class_precision: list[float]
    per_class_recall: list[float]
    per_class_f1: list[float]
    roc_auc: Optional[float]
    pr_auc: Optional[float]
    roc_fpr: Optional[list[float]]
    roc_tpr: Optional[list[float]]
    pr_precision: Optional[list[float]]
    pr_recall_pts: Optional[list[float]]
    status: str


class OCRResult(BaseModel):
    n_samples: int
    char_accuracy_pct: float
    word_accuracy_pct: float
    cer_pct: float
    wer_pct: float
    mean_confidence_pct: Optional[float]
    confidence_hist: Optional[list[float]]
    status: str


class RuntimeResult(BaseModel):
    n_samples: int
    mean_ms: float
    std_ms: float
    min_ms: float
    max_ms: float
    p50_ms: float
    p95_ms: float
    p99_ms: float
    fps: float
    target_fps: Optional[float]
    histogram: list[float]
    histogram_edges: list[float]
    status: str


class StabilityResult(BaseModel):
    n_runs: int
    pass_count: int
    fail_count: int
    pass_rate_pct: float
    failure_rate_pct: float
    first_failure_at: Optional[int]
    max_consecutive_failures: int
    fail_positions: list[int]
    trend: list[float]          # rolling pass rate (window=50)
    status: str


class ComponentScore(BaseModel):
    score: float
    weight: float
    label: str
    available: bool


class FinalScore(BaseModel):
    overall_score: float
    verdict: str                # PASS / CONDITIONAL PASS / FAIL
    component_scores: dict[str, ComponentScore]
    radar_labels: list[str]
    radar_values: list[float]
    recommendations: list[str]


class ValidationResult(BaseModel):
    system_name: str
    repeatability: Optional[RepeatabilityResult]
    accuracy: Optional[AccuracyResult]
    grr: Optional[GRRResult]
    ai: Optional[AIResult]
    ocr: Optional[OCRResult]
    runtime: Optional[RuntimeResult]
    stability: Optional[StabilityResult]
    final_score: FinalScore
    export_json: str
    export_csv: str

File: frontend/src/types/validation.ts
export interface GRRInput {
  values: number[][][]
  n_operators: number
  n_parts: number
  n_replicates: number
  tolerance?: number | null
}

export interface ValidationRequest {
  system_name?: string
  measurements?: number[]
  reference_value?: number | null
  usl?: number | null
  lsl?: number | null
  reference_values?: number[]
  predictions?: number[]
  ground_truth?: number[]
  scores?: number[]
  class_names?: string[]
  ocr_predicted?: string[]
  ocr_ground_truth?: string[]
  ocr_confidences?: number[]
  inference_times_ms?: number[]
  target_fps?: number | null
  stability_results?: boolean[]
  grr?: GRRInput | null
}

export interface RepeatabilityResult {
  n: number
  mean: number
  std: number
  min_val: number
  max_val: number
  range_val: number
  cp: number | null
  cpk: number | null
  cv_pct: number
  usl: number | null
  lsl: number | null
  histogram: number[]
  histogram_edges: number[]
  status: string
}

export interface AccuracyResult {
  n: number
  mae: number
  mape_pct: number
  bias: number
  rmse: number
  max_error: number
  errors: number[]
  status: string
}

export interface GRRResult {
  n_operators: number
  n_parts: number
  n_replicates: number
  ev_pct: number
  av_pct: number
  grr_pct: number
  pv_pct: number
  ndc: number
  var_repeatability: number
  var_reproducibility: number
  var_parts: number
  status: string
}

export interface AIResult {
  n_samples: number
  accuracy: number
  precision: number
  recall: number
  f1_score: number
  confusion_matrix: number[][]
  class_names: string[]
  per_class_precision: number[]
  per_class_recall: number[]
  per_class_f1: number[]
  roc_auc: number | null
  pr_auc: number | null
  roc_fpr: number[] | null
  roc_tpr: number[] | null
  pr_precision: number[] | null
  pr_recall_pts: number[] | null
  status: string
}

export interface OCRResult {
  n_samples: number
  char_accuracy_pct: number
  word_accuracy_pct: number
  cer_pct: number
  wer_pct: number
  mean_confidence_pct: number | null
  confidence_hist: number[] | null
  status: string
}

export interface RuntimeResult {
  n_samples: number
  mean_ms: number
  std_ms: number
  min_ms: number
  max_ms: number
  p50_ms: number
  p95_ms: number
  p99_ms: number
  fps: number
  target_fps: number | null
  histogram: number[]
  histogram_edges: number[]
  status: string
}

export interface StabilityResult {
  n_runs: number
  pass_count: number
  fail_count: number
  pass_rate_pct: number
  failure_rate_pct: number
  first_failure_at: number | null
  max_consecutive_failures: number
  fail_positions: number[]
  trend: number[]
  status: string
}

export interface ComponentScore {
  score: number
  weight: number
  label: string
  available: boolean
}

export interface FinalScore {
  overall_score: number
  verdict: string
  component_scores: Record<string, ComponentScore>
  radar_labels: string[]
  radar_values: number[]
  recommendations: string[]
}

export interface ValidationResult {
  system_name: string
  repeatability: RepeatabilityResult | null
  accuracy: AccuracyResult | null
  grr: GRRResult | null
  ai: AIResult | null
  ocr: OCRResult | null
  runtime: RuntimeResult | null
  stability: StabilityResult | null
  final_score: FinalScore
  export_json: string
  export_csv: string
}
